src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -858,6 +858,491 @@ uct_ib_mlx5_devx_dereg_keys(uct_ib_mlx5_md_t *md, uct_ib_mlx5_devx_mem_t *memh)
+    (*umr_mkey)->lkey |= UCT_IB_MLX5_MKEY_TAG_UMR;
+    (*umr_mkey)->rkey |= UCT_IB_MLX5_MKEY_TAG_UMR;
+
+    ucs_trace("%s: created UMR mkey %p lkey 0x%x",
+              uct_ib_device_name(&md->super.dev), *umr_mkey, (*umr_mkey)->lkey);
+
+    return UCS_OK;
+}
+
+static ucs_status_t
+uct_ib_mlx5_devx_umr_mkey_invalidate(uct_ib_mlx5_md_t *md,
+                                     uct_ib_mlx5_devx_umr_t *umr)
+{
+    uint32_t mkey         = umr->umr_mkey->lkey;
+    struct ibv_wc wc      = {};
+    struct ibv_send_wr wr = {
+        .wr_id           = 0,
+        .next            = NULL,
+        .num_sge         = 0,
+        .opcode          = IBV_WR_LOCAL_INV,
+        .send_flags      = IBV_SEND_INLINE | IBV_SEND_SIGNALED,
+        .invalidate_rkey = mkey
+    };
+    int ret;
+    struct ibv_send_wr *bad_wr;
+
+    ret = ibv_post_send(md->umr_qp, &wr, &bad_wr);
+    if (ret != 0) {
+        ucs_error("%s: failed to invalidate indirect mkey 0x%x: %m",
+                  uct_ib_device_name(&md->super.dev), mkey);
+        goto destroy_mkey;
+    }
+
+    /* Poll for completion */
+    while ((ret = ibv_poll_cq(md->umr_cq, 1, &wc)) == 0);
+
+    if ((ret < 0) || (wc.status != IBV_WC_SUCCESS)) {
+        ucs_error("%s: failed to poll CQ for invalidating indirect mkey 0x%x "
+                  "ret %d status %d", uct_ib_device_name(&md->super.dev), mkey,
+                  ret, wc.status);
+        goto destroy_mkey;
+    }
+
+    /* Add UMR mkey to mkey pool for reuse & cleanup */
+    ucs_list_add_head(&md->umr_mkey_pool, &umr->super);
+    ucs_trace("%s: put UMR mkey %p lkey 0x%x back to pool (%ld)",
+              uct_ib_device_name(&md->super.dev), umr->umr_mkey, mkey,
+              ucs_list_length(&md->umr_mkey_pool));
+    return UCS_OK;
+
+destroy_mkey:
+    if (mlx5dv_destroy_mkey(umr->umr_mkey) != 0) {
+        ucs_error("%s: failed to destroy UMR mkey %p lkey 0x%x: %m",
+                  uct_ib_device_name(&md->super.dev), umr->umr_mkey, mkey);
+    }
+    free(umr);
+    return UCS_ERR_IO_ERROR;
+}
+
+static uct_ib_mlx5_devx_umr_t *
+uct_ib_mlx5_devx_umr_mkey_pool_get(uct_ib_mlx5_md_t *md, int *from_pool)
+{
+    uct_ib_mlx5_devx_umr_t *umr;
+
+    *from_pool = !ucs_list_is_empty(&md->umr_mkey_pool);
+    if (*from_pool) {
+        umr = ucs_list_extract_head(&md->umr_mkey_pool, uct_ib_mlx5_devx_umr_t,
+                                    super);
+
+        ucs_trace("%s: extracted from pool UMR mkey %p lkey 0x%x",
+                  uct_ib_device_name(&md->super.dev), umr->umr_mkey,
+                  umr->umr_mkey->lkey);
+    } else {
+        umr = ucs_malloc(sizeof(*umr), "uct_ib_mlx5_devx_umr_t");
+        if (NULL == umr) {
+            ucs_error("%s: failed to allocate UMR mkey",
+                      uct_ib_device_name(&md->super.dev));
+            return NULL;
+        }
+
+        if (uct_ib_mlx5_devx_umr_mkey_create(md, &umr->umr_mkey) != UCS_OK) {
+            ucs_free(umr);
+            return NULL;
+        }
```

Comment:
maybe we can allow xgvmi access here when creating a new UMR? then we would not need `from_pool` var

Reply:
Ok, that looks good initially, but it's not that straightforward on practice.
The reason is that we have a fallback to umem registration in case xgvmi access can't be added:
```
    status = uct_ib_mlx5_devx_allow_xgvmi_access(md, memh, exported_lkey, 1);
    if (status == UCS_OK) {
        goto out;
    }

    ucs_debug("%s: indirect cross gvmi not supported, fallback to DEVX UMEM",
              uct_ib_device_name(&md->super.dev));
    mlx5dv_devx_obj_destroy(cross_mr);
    md->flags &= ~UCT_IB_MLX5_MD_FLAG_INDIRECT_XGVMI;

out_umem_mr:
    return uct_ib_mlx5_devx_xgvmi_umem_mr(md, memh);
```
Currently ig xgvmi fails once, then capability is removed and it falls back to umem_mr forever.
Now, if we try to allow xgvmi when creating UMR key, I'm afraid we can break that logic.
I will experiment a bit more, so far I didn't find a safe option to implement it like you suggested

Reply:
I believe there is no easy way to call `uct_ib_mlx5_devx_allow_xgvmi_access` during UMR creation, without changing existing behavior. There is one option however to get rid of `from_pool` variable: we can return `UCS_ERR_ALREADY_EXISTS` error code from the `bind` function when the resulting UMR mkey was present in the pool. Would it make sense to you?

Reply:
is it possible to try to allow xgvmi access during umr creation and update some cap flag on md if fail? So that we do not need to pass `from-pool` thru the whole stack, but check some flag instead?

Reply:
Ok, I think using some md flag is not a good option, for several reasons.
But elaborating this idea - there is better alternative actually! We can add `xgvmi_access` flag to the UMR mkey type itself:
```
typedef struct {
    ucs_list_link_t    super;
    struct mlx5dv_mkey *umr_mkey;
    /* Indicates whether XGVMI access had been allowed for this mkey */
    int     xgvmi_access;
} uct_ib_mlx5_devx_umr_t;
```
Sounds good?

Reply:
Ok, I implemented that, and I like the approach.
In principle, if we want to save a bit of memory, we can re-use mkey tag (8 LSB of mkey) for this purpose.
Assuming we remove `is_xgvmi_allowed` field, we can mark mkey as UMR only after allowing xgvmi access:
```
static void uct_ib_mlx5_devx_umr_mkey_set_xgvmi_allowed(uct_ib_mlx5_devx_umr_t *umr)
    /* Set marker for UMR key to distinguish it on the importer side */
    umr->umr_mkey->lkey  |= UCT_IB_MLX5_MKEY_TAG_UMR;
    umr->umr_mkey->rkey  |= UCT_IB_MLX5_MKEY_TAG_UMR;
}

// And to detect whether xgvmi is already set we can reuse existing function:
static UCS_F_ALWAYS_INLINE int uct_ib_mlx5_devx_is_umr_mkey(uint32_t lkey)
{
    return (UCT_IB_MLX5_MKEY_TAG_UMR == (lkey & 0xFF));
}

Reply:
I've just implemented and committed the aforementioned optimisation, please check